## 读取数据

导入数据分析所需要的库，并通过Pandas的 read_csv 函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为
DataFrame，并赋值给变量original_data。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv", encoding='utf-8', encoding_errors='replace')

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26:00       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26:00       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26:00       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26:00       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26:00       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的 original data这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行:结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

In [4]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
534828    581173     22737      RIBBON REEL CHRISTMAS PRESENT          5   
385804    570215     22469               HEART OF WICKER SMALL         6   
449244    575090     21166          COOK WITH WINE METAL SIGN          2   
364276    568654     22203              MILK PAN RED RETROSPOT         8   
420950    572913    85071D      CHARLIE+LOLA MY ROOM DOOR SIGN         5   
521192    580318     21936            RED RETROSPOT PICNIC BAG         5   
196232    553832     22434       BALLOON PUMP WITH 10 BALLOONS         8   
189977    553178     84949       SILVER HANGING T-LIGHT HOLDER         1   
378880    569669     22666     RECIPE BOX PANTRY YELLOW DESIGN         1   
270896    560601     21791  VINTAGE HEADS AND TAILS CARD GAME          1   

                InvoiceDate  UnitPrice  CustomerID         Country  
534828  12/07/2011 15:07:00       1.65     17870.0  United Kingdom  
385804  10/09/2011 11:53:00       1.65     16066.0  United Kingdom  
449244  11/08/2011 14:04:00       2.08     17515.0  United Kingdom  
364276  09/28/2011 12:20:00       3.75     14911.0            EIRE  
420950  10/26/2011 16:21:00       0.39     15993.0  United Kingdom  
521192  12/02/2011 15:39:00       2.95     17686.0  United Kingdom  
196232  05/19/2011 11:38:00       1.95     12957.0  United Kingdom  
189977  05/15/2011 11:29:00       1.65     17302.0  United Kingdom  
378880  10/05/2011 13:47:00       2.95     14730.0  United Kingdom  
270896  07/19/2011 17:17:00       1.25     16764.0  United Kingdom

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

### 评估数据干净程度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而Description、CustomerID 变量存在缺失值。

此外，InvoiceDate 的数据类型应为日期，CustomerID 的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失数据

在了解 Description 存在缺失值后，根据条件提取出缺失观察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity          InvoiceDate  \
622       536414     22139         NaN        56  12/01/2010 11:52:00   
1970      536545     21134         NaN         1  12/01/2010 14:32:00   
1971      536546     22145         NaN         1  12/01/2010 14:33:00   
1972      536547     37509         NaN         1  12/01/2010 14:33:00   
1987      536549    85226A         NaN         1  12/01/2010 14:34:00   
...          ...       ...         ...       ...                  ...   
535322    581199     84581         NaN        -2  12/07/2011 18:26:00   
535326    581203     23406         NaN        15  12/07/2011 18:31:00   
535332    581209     21620         NaN         6  12/07/2011 18:35:00   
536981    581234     72817         NaN        27  12/08/2011 10:33:00   
538554    581408     85175         NaN        20  12/08/2011 14:06:00   

        UnitPrice  CustomerID         Country  
622           0.0         NaN  United Kingdom  
1970          0.0         NaN  United Kingdom  
1971          0.0         NaN  United Kingdom  
1972          0.0         NaN  United Kingdom  
1987          0.0         NaN  United Kingdom  
...           ...         ...             ...  
535322        0.0         NaN  United Kingdom  
535326        0.0         NaN  United Kingdom  
535332        0.0         NaN  United Kingdom  
536981        0.0         NaN  United Kingdom  
538554        0.0         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失 Description 变量值。

从输出结果来看，这些缺失 Description 的交易数据，UnitPrice 都为0。为了验证猜想，我们增加筛选条件，看是否存在Description 变量缺失且 UnitPrice不为0的数据。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失 Description 值的数据，同时也不具备有效的 UnitPrice 值。

Description 表示产品名称， UnitPrice 表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

CustomerID 变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

                InvoiceDate  UnitPrice  CustomerID         Country  
622     12/01/2010 11:52:00       0.00         NaN  United Kingdom  
1443    12/01/2010 14:32:00       2.51         NaN  United Kingdom  
1444    12/01/2010 14:32:00       2.51         NaN  United Kingdom  
1445    12/01/2010 14:32:00       0.85         NaN  United Kingdom  
1446    12/01/2010 14:32:00       1.66         NaN  United Kingdom  
...                     ...        ...         ...             ...  
541536  12/09/2011 10:26:00       4.13         NaN  United Kingdom  
541537  12/09/2011 10:26:00       4.13         NaN  United Kingdom  
541538  12/09/2011 10:26:00       4.96         NaN  United Kingdom  
541539  12/09/2011 10:26:00      10.79         NaN  United Kingdom  
541540  12/09/2011 10:26:00    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID 表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些 customerID 缺失的交易数据仍然有
效，因此保留此变量为空的观察值。

#### 评估重复数据

根据数据变量的含义来看，虽然InvoiceNo、stockcode 和 customerID 都是唯一标识符，但一次交易可能包含多件商品，因此 InvoiceNo 可以存在重复，不同交易可以包含同件商品，因此 stockcode 可以存在重复，顾客可以进行多次交易或下单多个商品，因此 CustomerID 也可以存在重复。

那么针对此数据集，我们无需评估重复数据。

#### 评估不一致数据

不一致数据可能存在于 country 变量中，我们要查看是否存在多个不同值指代同一国家的情况

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


从 Country 变量值来看，“USA”、"United states”均在表示美国，"United Kingdom”、"UK"、
"U.K."均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

#### 评估无效或错误数据

可以通过DataFrame的 describe 方法，对数值统计信息进行快速了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看， Quantity 和 UnitPrice 存在负数，会对后续数值分析产生影响。
因此，我们先筛选出 Quantity 数值为负数的观察值，进一步评估其含义。

In [11]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

                InvoiceDate  UnitPrice  CustomerID         Country  
141     12/01/2010 09:41:00      27.50     14527.0  United Kingdom  
154     12/01/2010 09:49:00       4.65     15311.0  United Kingdom  
235     12/01/2010 10:24:00       1.65     17548.0  United Kingdom  
236     12/01/2010 10:24:00       0.29     17548.0  United Kingdom  
237     12/01/2010 10:24:00       0.29     17548.0  United Kingdom  
...                     ...        ...         ...             ...  
540449  12/09/2011 09:57:00       0.83     14397.0  United Kingdom  
541541  12/09/2011 10:28:00     224.69     15498.0  United Kingdom  
541715  12/09/2011 11:57:00      10.95     15311.0  United Kingdom  
541716  12/09/2011 11:58:00       1.25     17315.0  United Kingdom  
541717  12/09/2011 11:58:00       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，Quantity 变量为负数的观察值，InvoiceNo似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在 Quantity 变量为负数目 InvoiceNo 不以C开头的观察值。

In [13]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity          InvoiceDate  \
2406      536589     21777         NaN       -10  12/01/2010 16:50:00   
4347      536764    84952C         NaN       -38  12/02/2010 14:42:00   
7188      536996     22712         NaN       -20  12/03/2010 15:30:00   
7189      536997     22028         NaN       -20  12/03/2010 15:30:00   
7190      536998     85067         NaN        -6  12/03/2010 15:30:00   
...          ...       ...         ...       ...                  ...   
535333    581210     23395       check       -26  12/07/2011 18:36:00   
535335    581212     22578        lost     -1050  12/07/2011 18:38:00   
535336    581213     22576       check       -30  12/07/2011 18:38:00   
536908    581226     23090     missing      -338  12/08/2011 09:56:00   
538919    581422     23169     smashed      -235  12/08/2011 15:24:00   

        UnitPrice  CustomerID         Country  
2406          0.0         NaN  United Kingdom  
4347          0.0         NaN  United Kingdom  
7188          0.0         NaN  United Kingdom  
7189          0.0         NaN  United Kingdom  
7190          0.0         NaN  United Kingdom  
...           ...         ...             ...  
535333        0.0         NaN  United Kingdom  
535335        0.0         NaN  United Kingdom  
535336        0.0         NaN  United Kingdom  
536908        0.0         NaN  United Kingdom  
538919        0.0         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在 Quantity 变量为负数目 InvoiceNo 不以C开头的观察值。
但根据以上输出结果，这些筛选出的观察值的 Unitprice 观察值均为0，因此增加 Unitprice 的条件进行验证。

In [14]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当 Quantity 变量为负数的时候，观察值满足以下条件之一:

InvoiceNo 以C开头，表示订单被取消

UnitPrice为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将 Quantity 为负数的观察值删除。

接下来筛选出 UnitPrice 为负数的观察值。

In [15]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity          InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  08/12/2011 14:51:00   
299984   A563187         B  Adjust bad debt         1  08/12/2011 14:52:00   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看， UnitPrice 为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其鼎除。

## 清洗数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：

● 把 InvoiceDate 变量的数据类型转换为为日期时间。

● 把 CustomerID 变量的数据类型转换为字符串。

● 把 Description 变量缺失的观察值删除。

● 把 Country 变量值"USA"替换为"United States"。

● 把 Country 变量值"UK”、"U.K."替换为"United Kingdom"。

● 把 Quantity 变量值为负数的观察值删除。

● 把 UnitPrice 变量值为负数的观察值删除。

为了区分开经过清理的数据和原始的数据，我们创建新的变量 cleaned_data，让它为 original_data 复制出的副本。我们
之后的清理步骤都将被运用在 cleaned_data 上。

In [16]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26:00       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26:00       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26:00       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26:00       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26:00       3.39     17850.0  United Kingdom

把 InvoiceDate 变量的数据类型转换为为日期时间类型:

In [17]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把 CustomerID 变量的数据类型转换为字符串:

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值结尾的".0"删除：

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0,5)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把 Description 变量缺失的观察值删除，并查看删除后该列空缺值个数和:

In [22]:
cleaned_data.dropna(subset=["Description"], inplace=True)

In [23]:
cleaned_data["Description"].isnull().sum()

0

把 Country 变量值"USA"替换为"United States"，并检查替换后"USA"变量值个数：

In [24]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [25]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把 Country 变量值"UK"、"U.K."替换为"United xingdom"，并检查替换后"UK"和"U.K."变量值个数:

In [26]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [27]:
print(len(cleaned_data[cleaned_data["Country"] == "Uk"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.k."]))

0
0


把 Quantity 变量值为负数的观察值制除，并检査替换后 Quantity 变量值为负数的个数:

In [28]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [32]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把 UnitPrice 变量值为负数的观察值制除，并检査替换后 UnitPrice 变量值为负数的个数:

In [36]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [38]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv。

In [39]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom

In [40]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [42]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\ASUS\AppData\Local\Temp\ipykernel_25876\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom